### PR TITLE
[WIP] Proof of Concept CI_spellcheck

### PR DESCRIPTION
**Pull request by FeignInt**
Uses spellintian from Debian's Lintian Package
    
For testing purpose reverted: 973e1df044cd3e692a3852d33b9dc58c954c4f81
    
By default checks the compiled binary & Documents
    
    "*.ac" "*.md" "*.MD" "README*"
    "dosbox.1" "INSTALL"
    
i.e. Anything a typical end-user might encounter
:warning: list may be incomplete.
    
This test CI passes *--picky* , which flags up
capitalisation errors. It is usually incorrect
    
passing *--dosbox* to the script checks only the
*src/dosbox* binary
    
*--all* checks everything except translation files
Useful if wanting to clean up comments in source

### CODE REVIEW

**Comment by FeignInt** [comment]
[raw log](https://pipelines.actions.githubusercontent.com/RDKYnb9yFkXPfcXINdFWH3mzDaNqlgzrXM4kV8rnx6PJBbLLEF/_apis/pipelines/1/runs/24625/signedlogcontent/5?urlExpires=2020-12-08T17%3A16%3A11.8657407Z&urlSigningMethod=HMACV1&urlSignature=PaTP8nIAATFy91wLtVQtN9q%2Bcw1NA0%2F0pC73yhN1bTw%3D)

I forgot to mention, it gets the "typo"'s context 

```
2020-12-08T17:14:11.6443972Z {
2020-12-08T17:14:11.6444911Z   "filename": "BUILD.md",
2020-12-08T17:14:11.6445558Z   "typo": "depedencies",
2020-12-08T17:14:11.6446173Z   "fix": "dependencies",
2020-12-08T17:14:11.6446747Z   "context": [
2020-12-08T17:14:11.6447352Z     "BUILD.md:151:1. Install depedencies:"
2020-12-08T17:14:11.6447925Z   ]
2020-12-08T17:14:11.6448347Z }
2020-12-08T17:14:11.6448777Z {
2020-12-08T17:14:11.6449277Z   "filename": "BUILD.md",
2020-12-08T17:14:11.6449948Z   "typo": "python",
2020-12-08T17:14:11.6453615Z   "fix": "Python",
2020-12-08T17:14:11.6454080Z   "context": [
2020-12-08T17:14:11.6455484Z     "BUILD.md:399:  <https://github.com/andikleen/pmu-tools.> This is a collection of python"
2020-12-08T17:14:11.6456168Z   ]
2020-12-08T17:14:11.6456452Z }
2020-12-08T17:14:11.6456746Z {
2020-12-08T17:14:11.6457162Z   "filename": "CONTRIBUTING.md",
2020-12-08T17:14:11.6457689Z   "typo": "occassionally",
2020-12-08T17:14:11.6458158Z   "fix": "occasionally",
2020-12-08T17:14:11.6458571Z   "context": [
2020-12-08T17:14:11.6459267Z     "CONTRIBUTING.md:181:so occassionally you might find a commit originating from `svn/trunk` branch,"
2020-12-08T17:14:11.6460138Z   ]
2020-12-08T17:14:11.6460437Z }
2020-12-08T17:14:11.6460731Z {
2020-12-08T17:14:11.6461085Z   "filename": "INSTALL",
2020-12-08T17:14:11.6461492Z   "typo": "linux",
2020-12-08T17:14:11.6461844Z   "fix": "Linux",
2020-12-08T17:14:11.6462208Z   "context": [
2020-12-08T17:14:11.6462697Z     "INSTALL:53:    For ALSA audio support under linux."
2020-12-08T17:14:11.6463158Z   ]
2020-12-08T17:14:11.6463436Z }
2020-12-08T17:14:11.6463729Z {
2020-12-08T17:14:11.6464074Z   "filename": "README",
2020-12-08T17:14:11.6464484Z   "typo": "wether",
2020-12-08T17:14:11.6464862Z   "fix": "whether",
2020-12-08T17:14:11.6465238Z   "context": [
2020-12-08T17:14:11.6466044Z     "README:670:     - wether it can definitely not be changed at runtime"
2020-12-08T17:14:11.6466575Z   ]
2020-12-08T17:14:11.6467149Z }
2020-12-08T17:14:11.6467427Z {
2020-12-08T17:14:11.6467786Z   "filename": "README",
2020-12-08T17:14:11.6468166Z   "typo": "tcl",
2020-12-08T17:14:11.6468519Z   "fix": "Tcl",
2020-12-08T17:14:11.6468857Z   "context": [
2020-12-08T17:14:11.6469367Z     "README:463:        \"bps,spc,tcl,fcl\":",
2020-12-08T17:14:11.6469947Z     "README:467:           tcl: total clusters, between 1 and 65534",
2020-12-08T17:14:11.6470600Z     "README:468:           fcl: total free clusters, between 1 and tcl"
2020-12-08T17:14:11.6471071Z   ]
2020-12-08T17:14:11.6471375Z }
2020-12-08T17:14:11.6471655Z {
2020-12-08T17:14:11.6472018Z   "filename": "README",
2020-12-08T17:14:11.6472419Z   "typo": "russian",
2020-12-08T17:14:11.6472814Z   "fix": "Russian",
2020-12-08T17:14:11.6473190Z   "context": [
2020-12-08T17:14:11.6473744Z     "README:996:    2. To load one of russian keyboard layouts with codepage 866:",
2020-12-08T17:14:11.6474856Z     "README:998:       In order to type russian characters press ALT+RIGHT-SHIFT."
2020-12-08T17:14:11.6475442Z   ]
2020-12-08T17:14:11.6475739Z }
2020-12-08T17:14:11.6476020Z {
2020-12-08T17:14:11.6476383Z   "filename": "README",
2020-12-08T17:14:11.6476777Z   "typo": "french",
2020-12-08T17:14:11.6477161Z   "fix": "French",
2020-12-08T17:14:11.6477519Z   "context": [
2020-12-08T17:14:11.6478119Z     "README:999:    3. To load one of french keyboard layouts with codepage 850 (where the"
2020-12-08T17:14:11.6478667Z   ]
2020-12-08T17:14:11.6478964Z }
2020-12-08T17:14:11.6479245Z {
2020-12-08T17:14:11.6479870Z   "filename": "configure.ac",
2020-12-08T17:14:11.6480344Z   "typo": "libary",
2020-12-08T17:14:11.6480726Z   "fix": "library",
2020-12-08T17:14:11.6481086Z   "context": [
2020-12-08T17:14:11.6481959Z     "configure.ac:88:                   [AC_MSG_ERROR([Can't find a useable network libary])],"
2020-12-08T17:14:11.6482552Z   ]
2020-12-08T17:14:11.6482852Z }
2020-12-08T17:14:11.6483130Z {
2020-12-08T17:14:11.6483532Z   "filename": "configure.ac",
2020-12-08T17:14:11.6483975Z   "typo": "linux",
2020-12-08T17:14:11.6484344Z   "fix": "Linux",
2020-12-08T17:14:11.6484696Z   "context": [
2020-12-08T17:14:11.6485264Z     "configure.ac:605:dnl Check for mprotect. Needed for 64 bits linux",
2020-12-08T17:14:11.6486071Z     "configure.ac:647:    *-*-linux*)"
2020-12-08T17:14:11.6486477Z   ]
2020-12-08T17:14:11.6486771Z }
2020-12-08T17:14:11.6487051Z {
2020-12-08T17:14:11.6487458Z   "filename": "contrib/README.md",
2020-12-08T17:14:11.6487911Z   "typo": "absense",
2020-12-08T17:14:11.6488309Z   "fix": "absence",
2020-12-08T17:14:11.6488684Z   "context": [
2020-12-08T17:14:11.6489502Z     "contrib/README.md:12:  be used by `./configure` in the absense of `pkg-config`",
2020-12-08T17:14:11.6490474Z     "contrib/README.md:14: `./configure` in the absense of `pkg-config`",
2020-12-08T17:14:11.6491435Z     "contrib/README.md:16:  by `./configure` in the absense of `pkg-config`"
2020-12-08T17:14:11.6491969Z   ]
2020-12-08T17:14:11.6492266Z }
2020-12-08T17:14:11.6492549Z {
2020-12-08T17:14:11.6492979Z   "filename": "contrib/icons/icons.md",
2020-12-08T17:14:11.6493433Z   "typo": "Gnome",
2020-12-08T17:14:11.6493806Z   "fix": "GNOME",
2020-12-08T17:14:11.6494157Z   "context": [
2020-12-08T17:14:11.6494864Z     "contrib/icons/icons.md:42:Base design icon follows size and colour recommendations of Gnome HIG, so you"
2020-12-08T17:14:11.6495518Z   ]
2020-12-08T17:14:11.6495814Z }
2020-12-08T17:14:11.6496093Z {
2020-12-08T17:14:11.6496467Z   "filename": "src/dosbox",
2020-12-08T17:14:11.6496937Z   "typo": "implemenations",
2020-12-08T17:14:11.6497457Z   "fix": "implementations",
2020-12-08T17:14:11.6497886Z   "context": [
2020-12-08T17:14:11.6498223Z     ""
2020-12-08T17:14:11.6498508Z   ]
2020-12-08T17:14:11.6498802Z }
2020-12-08T17:14:11.6564966Z Post job cleanup.
```

File: scripts/verify-spelling-lite.sh [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  spellintian $picky "${file_list[@]/#/${repo_root}\/}" #| sed -E '/('"${filter}"')/d'
+}
+
+Get_context ()
+{
+  # get "offending" lines
+  jq -R -s '.[:-1]|split("\n";"")' < <(
+    # shellcheck disable=2046
+    git -C "${repo_root}" grep -n "${TYPO_FIX% ->*}" \
+        $( [[ $( file --mime-type "${repo_root}/${FILENAME%:}" ) =~ app.+exec ]] ||
```
Comment (FeignInt) [inline scripts/verify-spelling-lite.sh:47]: This seems to fail on github runner

it works locally

```json
{
  "filename": "src/dosbox",
  "typo": "implemenations",
  "fix": "implementations",
  "context": [
    "src/gui/render_crt_glsl.h:460:\tThese custom implemenations rely on the `rubyTextureSize` uniform variable."
  ]
}
```

Comment (kcgen) [reply scripts/verify-spelling-lite.sh:47]: Is `jq` installed and is it new enough? Maybe some shared MIME database `file` is using is older or not installed?

The default CI VM can be outdated (might need `ubuntu-latest`).

Comment (FeignInt) [reply scripts/verify-spelling-lite.sh:48]: yeah, it is probably the MIME db

I specifically targeted the ubuntu-20.04 in the linux.yml

The jq should be fine, it seems to be installed by default in the github runners

jq version could be sensitive, I'll worry about that when doing analysis, at the moment just filtering out "duplicate words"
as they are almost certainly false positives

there  a couple of false positives, which I was filtering ( turned off at the moment )

```
README: tcl -> Tcl
stb_vorbis.h: finalY -> finally 
```

Comment (FeignInt) [reply scripts/verify-spelling-lite.sh:48]:
```
[[ /home/runner/work/dosbox-staging/dosbox-staging/src/dosbox: application/x-sharedlib =~ app.+exec ]]
```
that is it

I get *application/x-pie-executable*

I guess I should do
```
[[ $foo =~ =~ app.+(x-pie-executable|x-sharedlib) ]]
```

**Comment by kcgen** [comment]
``` 
{
  "filename": "src/dosbox",
  "typo": "implemenations",
  "fix": "implementations",
  "context": [
    "src/gui/render_crt_glsl.h:460:\tThese custom implemenations rely on the `rubyTextureSize` uniform variable."
  ]
}
```

For the end-state output, can it be converted to YAML or just human-readable without the markup baggage? 

Or is the idea that these JSON chunks can be fed back into the tool, say to apply corrections or add them to an exclusion set (which is in JSON format)?

**Comment by FeignInt** [comment]
> ```
> {
>   "filename": "src/dosbox",
>   "typo": "implemenations",
>   "fix": "implementations",
>   "context": [
>     "src/gui/render_crt_glsl.h:460:\tThese custom implemenations rely on the `rubyTextureSize` uniform variable."
>   ]
> }
> ```
> 
> For the end-state output, can it be converted to YAML or just human-readable without the markup baggage?
> 
> Or is the idea that these JSON chunks can be fed back into the tool, say to apply corrections or add them to an exclusion set (which is in JSON format)?
> 
> Just curious. Not that the format is much of a bother.. but if there's room to tidy it up further, might as well :-)


is there a YAML equivalent of what **jq** is for JSON 

The JSON is just for filtering with *jq* later,
and I did make it human readable by not doing **jq -j -c** :wink: 

but yeah, .json , .yaml , or any machine readable output
```
script.sh > path/to/some.log
parselog.sh path/to/some.log
```
add have it print warnings / fails direct to the dashboard, so no need to dig around in downloaded zips or click through to raw text files

something I have wanted to look at but never got round to
[ debug , warning , error ](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#setting-a-debug-message)
 
looks simple enough
in fact, I can skip the whole *printf | jq* and just have *while read* spit out warnings; and error if it was src/dosbox

**Comment by kcgen** [comment]
> is there a YAML equivalent of what jq is for JSON

Yes: `yq`  ... 

``` shell
sudo apt install python3-setuptools
sudo pip3 install yq
```

then `cat <some JSON> | qy -y` ... and out come YAML.

But if you can do the fancy GitHub stuff, that might be nice too.

**Comment by FeignInt** [comment]
> > is there a YAML equivalent of what jq is for JSON
> 
> Yes: `yq` ...
> 
> ```shell
> sudo apt install python3-setuptools
> sudo pip3 install yq
> ```
> 
> then `cat <some JSON> | yq -y` ... and out comes YAML.
> 
Thanks, I'll have a look 
> The less ties into GitHub the better, so ideally we could launch this script locally and it would produce some nice output followed by a non-zero error code if spelling mistakes were found.
> 
There is actually two scripts, the first with half of it being nice to people and telling them how to install dependencies , even if they are not using Debian. (Lintian is Debian only, as far as I know)

I quickly shrunk to the -lite.sh , to see how well it works on github 

The difficult part is how to get the reason for failure passed to the PR...
I think it is time I tried reading the API docs again ( not the easiest to navigate )
It seems the most elegant solution is to post a report **here** , well not *here* but in the comments of the PR it just checked.

---

**SpellBot says NO**
Reason as follows:

filename: lineNO
blah typo blah
This **typo** should be **Correction** ?

---

Then just need a quick easy way to keep the exceptions up-to-date
at the moment I'm just building a string to use as a PATTERN

> For scripts that need to be "corrupted" with GitHub'isms, we have `.github/scripts` (if you need it).

Well, it didn't do what I expected it to do
just as above, it may still be githubified

File: scripts/verify-spelling-lite.sh [hunk 2]
```diff
@@ -115,6 +116,24 @@ _json_output ()
...
 )
 
+_dashboard_output ()
+{
+  while read -r FILENAME TYPO_FIX
+  do
+    FILENAME="${FILENAME#${repo_root}/}"
+    TYPO_FIX="${TYPO_FIX//\"}"
+
+    [[ ${TYPO_FIX} =~ (${filter}) ]] && continue
```
Comment (FeignInt) [inline scripts/verify-spelling-lite.sh:126]: should be
```
[[ "${FILENAME} ${TYPO_FIX}" =~ (${filter}) ]] && continue
```

**Comment by FeignInt** [comment]
Actually, the last commit did do something

It annotated  [the commit tab ](https://github.com/dosbox-staging/dosbox-staging/pull/763/commits/c61386a7d4ffc2e60eacf9a9069ea21d2171b8ee)

so instead of  echoing the "LINE" I should echo the typo/fix

**Comment by FeignInt** [comment]
Ok, I think this is working now:

- none-zero exit when src/dosbox has spelling error
- The Job Summary page has Annotations on it
- Files Changed has inline Annotations

I just need to clean it up and add documentation

Well, one other issue

At the moment, I have an incomplete list of ignores:

```bash
# The following patterns should be ignored
   filter="README: esc esc"
filter+="|README: tcl -> Tcl"
filter+="|stb_vorbis.h: finalY"
filter+="|duplicate word"
#TODO improve filter maintenance

[[ "${FILENAME} ${TYPO_FIX}" =~ (${filter}) ]] && continue
```

I'm thinking of putting the filter lines into a DOT file *.spellintian-ignore*

**Comment by FeignInt** [comment]
Added new feature

- --github-PR

Which I think should be the default
At the moment it is just at warning level, but I can change it to exit 1+ 

the github PR Mode gets the list of files changed via github's api
I guess I could re-write it to get this info from git, but it just seemed teh easiest way to do it

For this  test I added to analysis.yml,  with

    --picky --github-PR --docs

- *--picky* be pedantic about capitalisation, those will never result in none-zero exit
- *--docs* is to enforce the 'old' default file list, without it it will only look at the files changed by the PR

The --docs isn't really needed, just for demonstration here

I will probably make github-PR default ( based on running in a gh runner )
and the --picky will never result in  a fatal error, just a warning so might as well make that default.

see https://github.com/dosbox-staging/dosbox-staging/pull/763/files
for annotation (have to scroll down a bit)

**Comment by FeignInt** [comment]
Spellintian may lack an extensive word list

https://github.com/dosbox-staging/dosbox-staging/pull/769/files#r542992059

so I need to seek something to fill that gap

**Comment by dreamer** [comment]
@FeignInt there's a lot of good work in here - since this is still a draft I won't go into details, but I'll leave some suggestions:

- bash script kind-of grew already… and bash scripts that grow fast tend to end up being write-once scripts. Personally, as a rule of thumb, when bash script reaches 100 lines I focus on rewriting it in python (following Google's code style guide). Several scripts in this repo went through this already (e.g. `scripts/count-*`), and for others - I wish they did (e.g. `scripts/import-from-svn/*` or `format-commit.sh`). Therefore I suggest rewriting in python while it's still fresh in your memory (because whenever I let my long bash script grow, it always bit me after some time). It's only a suggestion though :)

- Long time ago me and @kcgen had discussion about formatting for bash scripts and I think we ended up deciding on using tabs for indentation (as we already use tabs in C/C++ code). Having a script with 2-space indentation looks somewhat weird to me. Again, not a big deal for adjunct scripts (we have restrictive formatting rules for C/C++, but for scripting we can stray away somewhat). For python we use 4 spaces instead of tabs and I would prefer to stick to this.

- Initially I imagined this will be much simpler script - just running lintian as lightweight checker in on in `.github/workflows/analysis.yml` - and if it fails, it fails. But your work on injecting comments into reviews is very interesting :)

**Comment by FeignInt** [comment]
This recent bug is just a distraction now
```
++ sed -E -n '457,+7{s/^([0-9]+):.*\bimplemenations\b.*/\1/p}'
2020-12-18T08:53:35.3540659Z ++ git -C /home/runner/work/dosbox-staging/dosbox-staging grep -h -n '' -- src/gui/render_crt_glsl.h
2020-12-18T08:53:35.3605410Z ##[warning]src/gui/render_crt_glsl.h: implemenations -> implementations
2020-12-18T08:53:35.4030172Z ++ git -C /home/runner/work/dosbox-staging/dosbox-staging grep -h -n '' -- .github/scripts/verify-spelling-spellintian.sh
2020-12-18T08:53:35.4031538Z ++ sed -E -n '\";\"\")"]{s/^([0-9]+):.*\bpython\b.*/\1/p}'
2020-12-18T08:53:35.4045747Z sed: -e expression #1, char 9: Unmatched ) or \)
```

seems to be some quirk with versions, locally on Debian Sid the script never gets that far in the while loop

I'm killing off spellintian, as codespell seems to be a superior alternative

Makes sense to close this PR, and I'll reopen a new one specific to [codespell](https://github.com/codespell-project/codespell)

**Pull request closed without merging**